File: src/core/inventory.ts
/**
 * La despensa.
 *
 * Hasta acá la comida era infinita: `alimentar()` no consultaba nada y podías
 * dar de comer cien veces seguidas. El principio de "toda acción cuesta algo"
 * estaba implementado a nivel de estadísticas —comer de más hace mal— pero no
 * a nivel de economía, que es donde falta la otra mitad.
 *
 * El inventario es del jugador, no de la criatura: es una despensa, no un
 * estómago. Por eso vive en el estado de la partida y no en cada criatura.
 */

import type { FoodKind } from "./evolution.ts";

export type Inventario = Record<string, number>;

/**
 * Con qué se arranca.
 *
 * Alcanza para los primeros días sin tener que salir de expedición enseguida.
 * El cristal empieza en cero a propósito: es lo raro, y tiene que sentirse así.
 */
export function inventarioInicial(): Inventario {
  return { baya: 3, raiz: 2, larva: 2, cristal: 0 };
}

export function cuanto(inventario: Inventario, id: string): number {
  return inventario[id] ?? 0;
}

export function hay(inventario: Inventario, id: string): boolean {
  return cuanto(inventario, id) > 0;
}

export function total(inventario: Inventario): number {
  return Object.values(inventario).reduce((suma, n) => suma + n, 0);
}

/**
 * Descuenta una unidad. Devuelve `null` si no había.
 *
 * No muta el inventario que recibe: devolver uno nuevo permite que quien llama
 * decida si el cambio se aplica, y evita descontar comida cuando la acción que
 * venía después terminó fallando.
 */
export function consumir(inventario: Inventario, id: string): Inventario | null {
  if (!hay(inventario, id)) return null;
  return { ...inventario, [id]: cuanto(inventario, id) - 1 };
}

export function agregar(inventario: Inventario, id: string, cantidad = 1): Inventario {
  if (cantidad <= 0) return { ...inventario };
  return { ...inventario, [id]: cuanto(inventario, id) + cantidad };
}

/** Suma un botín entero de una vez. */
export function agregarVarios(inventario: Inventario, botin: Inventario): Inventario {
  let resultado = { ...inventario };
  for (const [id, cantidad] of Object.entries(botin)) {
    resultado = agregar(resultado, id, cantidad);
  }
  return resultado;
}

/** Qué tipo de comida es cada id. Lo usa el botín de las expediciones. */
export const TIPO_POR_ALIMENTO: Record<string, FoodKind> = {
  baya: "dulce",
  raiz: "mineral",
  larva: "proteina",
  cristal: "raro",
};

File: gdext/src/inventory.cpp
/**
 * inventory.cpp — ver inventory.h.
 */

#include "inventory.h"

namespace petbits {

int64_t Inventario::cuanto(std::string_view id) const {
    for (const auto& [k, v] : items_) {
        if (k == id) return v;
    }
    return 0;
}

bool Inventario::hay(std::string_view id) const {
    return cuanto(id) > 0;
}

int64_t Inventario::total() const {
    int64_t suma = 0;
    for (const auto& [k, v] : items_) suma += v;
    return suma;
}

void Inventario::poner(std::string_view id, int64_t cantidad) {
    for (auto& [k, v] : items_) {
        if (k == id) {
            v = cantidad;
            return;
        }
    }
    items_.emplace_back(std::string(id), cantidad);
}

bool Inventario::consumir(std::string_view id) {
    if (!hay(id)) return false;
    poner(id, cuanto(id) - 1);
    return true;
}

void Inventario::agregar(std::string_view id, int64_t cantidad) {
    if (cantidad <= 0) return;
    poner(id, cuanto(id) + cantidad);
}

Inventario inventarioInicial() {
    Inventario i;
    // El mismo orden que el objeto literal del TS, para que el JSON salga igual.
    i.poner("baya", 3);
    i.poner("raiz", 2);
    i.poner("larva", 2);
    i.poner("cristal", 0);
    return i;
}

bool tipoDeAlimento(std::string_view id, FoodKind& salida) {
    if (id == "baya") { salida = FoodKind::Dulce; return true; }
    if (id == "raiz") { salida = FoodKind::Mineral; return true; }
    if (id == "larva") { salida = FoodKind::Proteina; return true; }
    if (id == "cristal") { salida = FoodKind::Raro; return true; }
    return false;
}

} // namespace petbits
